### ISSUE TITLE
[V2] TimeoutException with DeviceClient.OpenAsync

### ISSUE THREAD

**Issue by leeprtc** [Azure/iot-edge#511]
What would cause a TimeoutException with DeviceClient.OpenAsync()? Is it common to see this exception?

I've seen this a few times now. I'm running with a few (2-3) custom modules with one module sending a message every 5 or 10 seconds and the other modules doing something with these messages.

**Comment by varunpuranik** [comment]
Can you provide a little more information on your setup and scenario? Are the deployed modules seeing this exception, or is it some downstream devices? Does it happen randomly, or with specific devices/modules? What hardware are you running your scenario on?

Assuming that this happens randomly to one of the deployed modules, you should typically not see this happen. Is the EdgeHub able to start properly? Can you look at the EdgeHub logs when you get a repro and copy/paste them here? 

We recommend adding retry logic to all your operations in case EdgeHub takes longer to boot or re-starts for some reason.

**Comment by leeprtc** [comment]
I was running with Windows container on Windows 10, using Docker for Windows. I had one module generating some data every 5 or 10 seconds and then the other modules doing something with the data. It is the deployed modules that is seeing this exception. As far as I could tell, this exception happens randomly. I have not been able to determine a consistent repro.

In the process of investigation, I stumbled upon and narrow down the issue in #515. I'm suspecting maybe issue #515 also caused this issue.

If I managed to repro, I will copy the EdgeHub logs here.

**Comment by leeprtc** [comment]
I repro the issue. It seems easy to repro. I added two modules. One module generate some data every 10 second. The other module receives the data, print it, and then forward the data to Azure.

Repro steps
1. Add the two modules - around 16:36:23
2. Remove one module (that one that prints data and then forward to Azure) - around 16:38:04
3. Add the module in step 2 back - around 16:39:56

The module did not see the TimeoutException until around 16:44:16. 

Looking at the EdgeHub logs, it seems like after step 3, EdgeHub gets a bunch of exceptions and keeps retrying, which finally led to the TimeoutException. (I changed my module to add retry logic so, near the end, you will see another attempt to call OpenAsync.)

EdgeHub logs attached.

[openasync_timeout.log](https://github.com/Azure/iot-edge/files/1787305/openasync_timeout.log)

**Comment by varunpuranik** [comment]
Some of the errors in the logs are not expected, so we will look into this issue. 
Can you please tell me how you are removing and adding the module in steps 2 and 3? Are you doing a SetModules operation via the Azure Portal?

**Comment by leeprtc** [comment]
Yes, I'm using SetModules operation in Azure Portal to remove the module.
